code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/LoadBalancerServiceActivatePreListener.java

```diff
@@ -0,0 +1,50 @@
+package io.cattle.platform.servicediscovery.process;
+
+import static io.cattle.platform.core.model.tables.LoadBalancerTable.LOAD_BALANCER;
+import io.cattle.platform.core.constants.CommonStatesConstants;
+import io.cattle.platform.core.model.LoadBalancer;
+import io.cattle.platform.core.model.Service;
+import io.cattle.platform.engine.handler.HandlerResult;
+import io.cattle.platform.engine.handler.ProcessPreListener;
+import io.cattle.platform.engine.process.ProcessInstance;
+import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.object.process.StandardProcess;
+import io.cattle.platform.process.common.handler.AbstractObjectProcessLogic;
+import io.cattle.platform.servicediscovery.api.constants.ServiceDiscoveryConstants;
+import io.cattle.platform.util.type.Priority;
+
+import javax.inject.Named;
+
+@Named
+public class LoadBalancerServiceActivatePreListener extends AbstractObjectProcessLogic implements ProcessPreListener,
+        Priority {
+
+    @Override
+    public String[] getProcessNames() {
+        return new String[] { ServiceDiscoveryConstants.PROCESS_SERVICE_ACTIVATE };
+    }
+
+    @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        Service service = (Service) state.getResource();
+
+        if (!service.getKind().equalsIgnoreCase(ServiceDiscoveryConstants.KIND.LOADBALANCERSERVICE.name())) {
+            return null;
+        }
+
+        LoadBalancer lb = objectManager.findOne(LoadBalancer.class, LOAD_BALANCER.SERVICE_ID, service.getId(),
+                LOAD_BALANCER.REMOVED, null);
+        if (lb != null
+                && (lb.getState().equalsIgnoreCase(CommonStatesConstants.INACTIVE) || lb.getState().equalsIgnoreCase(
```

Comment:
Its done this way to handle idempotent retries for "execute". Otherwise the first failed attempt will change the state to Activating, but won't proceed further, so the LB will get stuck forever in Activating state.